torch/fx/_graph_pickler.py

```diff
@@ -0,0 +1,582 @@
+
+            return self._getattr_by_name(math, self.name)
+        elif self.root == "torch":
+            return self._getattr_by_name(torch, self.name)
+        else:
+            raise NotImplementedError
+
+
+class _OpOperatorPickleData(_OpPickleData):
+    def __init__(self, name: str) -> None:
+        self.name = name
+
+    def unpickle(self, unpickle_state: _UnpickleState) -> object:
+        import operator
+
+        return self._getattr_by_name(operator, self.name)
+
+
+class _GraphPickleData:
+    def __init__(self, graph: torch.fx.Graph) -> None:
+        self.tracer_cls = graph._tracer_cls
+        self.tracer_extras = graph._tracer_extras
+
+        nodes: Dict[torch.fx.Node, _NodePickleData] = {}
+        for node in graph.nodes:
+            nodes[node] = _NodePickleData(node, nodes)
+        self.nodes = tuple(nodes.values())
+
+        # Unpickled variables:
+        # self._used_names = graph._used_names
+        # -- self._insert = self._root.prepend
+        # self._len = graph._len
+        # self._graph_namespace = graph._graph_namespace
+        # self._owning_module = graph._owning_module
+        # self._codegen = graph._codegen
+        # self._co_fields: Dict[str, Any] = graph._co_fields
+        # -- self._find_nodes_lookup_table = _FindNodesLookupTable()
+
+    def unpickle(
+        self, gm: torch.fx.GraphModule, unpickle_state: _UnpickleState
+    ) -> torch.fx.Graph:
+        graph = torch.fx.Graph(gm, self.tracer_cls, self.tracer_extras)
+
+        nodes: Dict[_NodePickleData, torch.fx.Node] = {}
+        for nd in self.nodes:
+            nodes[nd] = nd.unpickle(graph, nodes, unpickle_state)
+
+        return graph
+
+
+class _TracingContextPickleData:
+    @classmethod
+    def reduce_helper(
+        cls, pickler: GraphPickler, obj: torch._guards.TracingContext
+    ) -> Tuple[
+        Callable[[Self, _UnpickleState], torch._guards.TracingContext],
+        Tuple[Self, _UnpickleStateToken],
+    ]:
+        return (
+            cls.unpickle,
+            (
+                cls(obj),
+                pickler._unpickle_state,
+            ),
+        )
+
+    def __init__(self, context: TracingContext) -> None:
+        # TODO: Do we really need all of this?
+        self.module_context = context.module_context
+        self.frame_summary_stack = context.frame_summary_stack
+        self.loc_in_frame = context.loc_in_frame
+        self.aot_graph_name = context.aot_graph_name
+        self.params_flat = context.params_flat
+        self.params_flat_unwrap_subclasses = context.params_flat_unwrap_subclasses
+        self.params_unwrapped_to_flat_index = context.params_unwrapped_to_flat_index
+        self.output_strides = context.output_strides
+        self.force_unspec_int_unbacked_size_like = (
+            context.force_unspec_int_unbacked_size_like
+        )
+        # Not saved (because it's difficult and maybe not needed?):
+        #   self.fw_metadata = context.fw_metadata
```

Comment:
It's hard for me to reason about the intentions here. Is the goal that when unpickling a graph module that we also reinstate the exact tracing context that it came from?

A lot of these seem specific to the tracing step that uses the graph module, instead of being "owned" by the graph module itself. Do you have a sense of how the GraphModule uses the fields in this tracingcontext (At least, the ones that you do end up pickling and saving)? 

I'm just a bit fuzzy on what exactly this context is doing

Reply:
The intentions were that I just wanted to save what was needed and no more. (There's a lot of "junk" hanging off some of our structures). I listed what I didn't save so in case we're having a problem and grep for it we'll see that it's explicitly not saved (and maybe that should be fixed).